lib/metrics/context.js

```diff
@@ -6,42 +6,145 @@
-  }).and('flowId', 'flowBeginTime').optional(),
-
-  add: function (data, metadata, doNotTrack) {
-    if (metadata) {
-      data.time = Date.now()
-      data.flow_id = metadata.flowId
-      data.flow_time = calculateFlowTime(data.time, metadata.flowBeginTime)
-      data.context = metadata.context
-      data.entrypoint = metadata.entrypoint
-      data.migration = metadata.migration
-      data.service = metadata.service
-
-      if (! doNotTrack) {
-        data.utm_campaign = metadata.utmCampaign
-        data.utm_content = metadata.utmContent
-        data.utm_medium = metadata.utmMedium
-        data.utm_source = metadata.utmSource
-        data.utm_term = metadata.utmTerm
+var SCHEMA = isA.object({
+  flowId: isA.string().length(64).regex(HEX).optional(),
+  flowBeginTime: isA.number().integer().positive().optional(),
+  context: isA.string().optional(),
+  entrypoint: isA.string().optional(),
+  migration: isA.string().optional(),
+  service: isA.string().optional(),
+  utmCampaign: isA.string().optional(),
+  utmContent: isA.string().optional(),
+  utmMedium: isA.string().optional(),
+  utmSource: isA.string().optional(),
+  utmTerm: isA.string().optional()
+}).and('flowId', 'flowBeginTime').optional()
+
+module.exports = function (log, config) {
+  var _memcached
+
+  return {
+    schema: SCHEMA,
+    save: save,
+    copy: copy,
+    remove: remove
+  }
+
+  /**
+   * Saves metrics context metadata against the current session.
+   *
+   * @param sessionToken session token
+   * @param metadata     metrics context metadata
+   */
+  function save (sessionToken, metadata) {
+    var memcached = getMemcached()
+    return memcached.setAsync(sessionToken.tokenId.toString('hex'), metadata, 0)
+      .catch(function (err) {
+        log.error({ op: 'metricsContext.save', err: err })
+      })
+  }
+
+  function getMemcached () {
+    try {
+      if (! _memcached) {
+        _memcached = new Memcached(config.memcache.address, {
+          timeout: 500,
+          retries: 1,
+          retry: 1000,
+          reconnect: 1000,
+          idle: 30000,
+          namespace: 'fxa-metrics~'
+        })
+      }
+
+      return _memcached
+    } catch (err) {
+      log.error({ op: 'metricsContext.getMemcached', err: err })
+
+      // If memcached is missing, turn its methods into nops
+      return {
+        delAsync: nop,
+        getAsync: nop,
+        setAsync: nop
       }
     }
+  }
+
+  function nop () {
+    return P.resolve()
+  }
+
+  /**
+   * Copies metrics context metadata to data, using either metadata
+   * passed in from a request or previously-saved metadata for the
+   * current session.
+   *
+   * @param data         target object
+   * @param metadata     metrics context metadata
+   * @param sessionToken session token
+   * @param doNotTrack   flag indicating whether DNT request header
+   *                     was specified
+   */
+  function copy (data, metadata, sessionToken, doNotTrack) {
```

Comment:
I like how this has come together, hiding the complexity of retrieving the data behind a simple "log the context for this session token" interface.